src/expr/src/scalar/mod.rs

```diff
@@ -890,6 +897,114 @@ impl MirScalarExpr {
                 }
             },
         );
+
+        /* #region `reduce_list_create_list_index_literal` and helper functions */
+
+        fn list_create_type(list_create: &MirScalarExpr) -> ScalarType {
+            if let MirScalarExpr::CallVariadic {
+                func: VariadicFunc::ListCreate { elem_type: typ },
+                ..
+            } = list_create
+            {
+                (*typ).clone()
+            } else {
+                unreachable!()
+            }
+        }
+
+        fn is_list_create_call(expr: &MirScalarExpr) -> bool {
+            matches!(
+                expr,
+                MirScalarExpr::CallVariadic {
+                    func: VariadicFunc::ListCreate { .. },
+                    ..
+                }
+            )
+        }
+
+        fn reduce_list_create_list_index_literal(
+            top_list_create: &mut MirScalarExpr,
+            index_exprs: &mut Vec<MirScalarExpr>,
+        ) -> MirScalarExpr {
+            // E.g.: LIST[f1,f2][2] --> f2. We handle multi-dimensional lists as well, e.g.
+            // LIST[[[f1, f2], [f3, f4]], [[f5, f6], [f7, f8]]] [2][n][2] --> LIST[f6, f8] [n]
+            // We iterate over the index_exprs and remove literals, but keep non-literals.
+            // When we encounter a non-literal, we need to dig into the nested ListCreates:
+            // `list_creates` will contain all the ListCreates of the current level. If an element of
+            // `list_creates` is not actually a ListCreate, then we break out of the loop.
+            // When we remove a literal, we need to partial-evaluate all ListCreates
+            // that are at the current level (except those that disappeared due to
+            // literals at earlier levels), and change each element in `list_creates`
+            // to the result of the partial evaluation.
+            let top_list_create_mv = &mut (*top_list_create);
+            let mut list_creates = vec![top_list_create_mv];
+            let mut i = 0;
+            while i < index_exprs.len() && list_creates.iter().all(|lc| is_list_create_call(lc)) {
+                if index_exprs[i].is_literal_ok() {
+                    // We can remove this index.
+                    let removed_index = index_exprs.remove(i);
+                    let index_i64 = match removed_index.as_literal().unwrap().unwrap() {
+                        Datum::Int64(sql_index_i64) => sql_index_i64 - 1,
+                        _ => unreachable!(), // always an Int64, see plan_index_list
+                    };
+                    // substitute each list_create in list_creates by its `index`th argument (or null)
```

Comment:
What is happening here is that for each cursor, we modify each subexpression by substituting it with its `index`th argument (or null).